### PR TITLE
[hotfix][unwind_native] optimize the instruction count of the unwind_…

### PR DESCRIPTION
**Pull request by tsint**
Older kernels (such as 4.19) do not allow the instruction count of BPF programs to exceed 4096. However, as shown below, the latest version of the unwind_native program already has 4139 instructions.
```
.text has 0 instructions
perf_event/unwind_dotnet has 3591 instructions
perf_event/unwind_hotspot has 3073 instructions
tracepoint/sched/sched_switch has 1155 instructions
tracepoint/syscalls/sys_enter_read has 22 instructions
perf_event/unwind_stop has 445 instructions
perf_event/native_tracer_entry has 432 instructions
perf_event/unwind_native has 4139 instructions
perf_event/unwind_perl has 2777 instructions
perf_event/unwind_php has 2583 instructions
perf_event/unwind_python has 3778 instructions
perf_event/unwind_ruby has 3258 instructions
tracepoint/sched/sched_process_exit has 111 instructions
tracepoint/syscalls/sys_enter_bpf has 25 instructions
raw_tracepoint/sys_enter has 29 instructions
perf_event/unwind_v8 has 3337 instructions

Total instructions: 28755
```
Without modifying the logic of the unwind_native program, I optimized the instruction count of the unwind_native program to fit within the 4096 limit.

### CODE REVIEW

**Comment by florianl** [comment]
hi @tsint 
And thank you for your contribution! Can you provide some information on how you compiled the eBPF programs that resulted in these increased instruction counts?

When checking the instruction count for the checked in eBPF blobs, we get the following at the moment:
```
$ ./print_instruction_count.sh tracer.ebpf.amd64 
/usr/bin/llvm-objdump

Instruction counts for tracer.ebpf.amd64:

.text has 0 instructions
perf_event/unwind_dotnet has 3597 instructions
perf_event/unwind_hotspot has 3073 instructions
tracepoint/sched/sched_switch has 1160 instructions
tracepoint/syscalls/sys_enter_read has 22 instructions
perf_event/unwind_stop has 445 instructions
perf_event/native_tracer_entry has 431 instructions
perf_event/unwind_native has 3995 instructions
perf_event/unwind_perl has 2646 instructions
perf_event/unwind_php has 2563 instructions
perf_event/unwind_python has 3739 instructions
perf_event/unwind_ruby has 3255 instructions
tracepoint/sched/sched_process_exit has 111 instructions
tracepoint/syscalls/sys_enter_bpf has 25 instructions
raw_tracepoint/sys_enter has 29 instructions
perf_event/unwind_v8 has 3317 instructions

Total instructions: 28408
```

With these eBPF blobs we do not have eBPF programs with more than 4096 instructions. There is also [a CI check in place](https://github.com/open-telemetry/opentelemetry-ebpf-profiler/blob/main/.github/workflows/unit-test-on-pull-request.yml#L78), to make sure this limit of instructions is respected and eBPF programs load on Linux kernel 4.19 - [see example run here.](https://github.com/open-telemetry/opentelemetry-ebpf-profiler/actions/runs/10493825873/job/29068780034).

**Comment by tsint** [comment]
Sorry, this might be because I used llvm-14 for compilation, which caused the generated BPF program to exceed 4096 instructions.

**Comment by rockdaboot** [comment]
Thank you @tsint
Can you sign the CLA and add the two ebpf binaries built with clang-16 to this PR?
```
support/ebpf/tracer.ebpf.amd64
support/ebpf/tracer.ebpf.arm64
```
If that is a problem, let us know and we take care.

@florianl With clang-16, this PR shaves off some instructions:
amd64: 3995 -> 3971
arm64: 4060 -> 4025

**Comment by tsint** [comment]
@rockdaboot 
OK, I'll try compiling and building with clang-16. Please wait for my test results.

**Comment by tsint** [comment]
@rockdaboot 
Hello, I recompiled using clang-16 and have added two EBPF binary files to this PR.

**Review by rockdaboot — APPROVED** [review]
Thanks, LGTM

We have to improve the CI yet to automatically build/check the ebpf blobs :)

File: support/ebpf/native_stack_trace.ebpf.c [hunk 1]
```diff
@@ -397,9 +397,9 @@ static ErrorCode unwind_one_frame(u64 pid, u32 frame_idx, UnwindState *state, bo
 
   // The relevant executable is compiled with frame pointer omission, so
   // stack deltas need to be retrieved from the relevant map.
-  ErrorCode error = get_stack_delta(state->text_section_id, state->text_section_offset,
-                                    &addrDiff, &unwindInfo);
+  ErrorCode error = get_stack_delta(state, &addrDiff, &unwindInfo);
   if (error) {
+    increment_metric(state->error_metric);
```
Comment (fabled) [inline support/ebpf/native_stack_trace.ebpf.c:402]: This should be removed. The error path ends up going to `unwind_stop` which does the increment also at https://github.com/open-telemetry/opentelemetry-ebpf-profiler/blob/main/support/ebpf/interpreter_dispatcher.ebpf.c#L210

File: support/ebpf/native_stack_trace.ebpf.c [hunk 2]
```diff
@@ -489,9 +489,9 @@ static ErrorCode unwind_one_frame(u64 pid, u32 frame_idx, struct UnwindState *st
...
```
Comment (fabled) [inline support/ebpf/native_stack_trace.ebpf.c:494]: Same here. The error path handler does automatic `error_metric` incremenet.

**Review by fabled — CHANGES_REQUESTED** [review]
Simplification looks great thanks! The error path metric incrementing needs adjustment and probably that shaves even more bytes out.

**Comment by tsint** [comment]
@fabled 
OK, I have removed those two lines of code in the PR based on your feedback. Thanks!

**Review by christos68k — APPROVED** [review]

**Review by fabled — APPROVED** [review]
LGTM

**Pull request merged**